CI failure (`Python fast`) is pre-existing and unrelated to this PR.

Failing test: `python/tests/test_dae_fd_mol.py::TestMOLCollocation::test_mol_collocation_solves`

```
TypeError: Gradient only defined for scalar-output functions. Output had shape: (3, 3).
python/discopt/_jax/problem_classifier.py:951: TypeError
```

The bug is in `extract_lp_data` (line 951), which calls `jax.grad(fn)(x_zero)` on each constraint function — but vector-valued constraint bodies (DAE collocation, MOL) have non-scalar output and `jax.grad` requires scalar output.

Verified to reproduce on `main` with no PR changes:

```
$ git checkout main
$ pytest python/tests/test_dae_fd_mol.py::TestMOLCollocation::test_mol_collocation_solves
FAILED python/tests/test_dae_fd_mol.py::TestMOLCollocation::test_mol_collocation_solves
TypeError: Gradient only defined for scalar-output functions. Output had shape: (3, 3).
```

Same root-cause hits `test_dae.py::TestCollocationODE::test_exponential_decay` locally (different vector shape `(10, 3)`).

This PR's diff touches only `solvers/`, `solver.py`, `modeling/core.py`, `validation/examiner.py`, and `tests/test_examiner.py` — none of which interact with `problem_classifier.extract_lp_data`. All 17 examiner tests pass. Tracking the LP-extractor bug separately.

--- a/python/discopt/modeling/core.py
+++ b/python/discopt/modeling/core.py
@@ -865,6 +865,17 @@ class SolveResult:
     jax_time: float = 0.0
     python_time: float = 0.0
 
+    # KKT duals at the returned point, when the underlying solver exposes them.
+    # ``constraint_duals`` is keyed by Constraint.name; entries with a vector
+    # body have one multiplier per row. ``bound_duals_lower`` /
+    # ``bound_duals_upper`` are keyed by Variable.name. All values are in the
+    # internal-minimization sign convention (``>= 0`` at active bounds /
+    # binding-from-below inequalities). For maximize problems, the multipliers
+    # correspond to the negated objective the solver actually saw.
+    constraint_duals: Optional[dict[str, np.ndarray]] = None
+    bound_duals_lower: Optional[dict[str, np.ndarray]] = None
+    bound_duals_upper: Optional[dict[str, np.ndarray]] = None
+
     # Convex fast path indicator
     convex_fast_path: bool = False
 

--- a/python/discopt/solver.py
+++ b/python/discopt/solver.py
@@ -977,6 +977,54 @@ def _unpack_solution(model: Model, x_flat: np.ndarray):
     return result
 
 
+def _unpack_constraint_duals(
+    evaluator, mult_g: Optional[np.ndarray]
+) -> Optional[dict[str, np.ndarray]]:
+    """Slice a flat constraint-multiplier vector into a dict keyed by
+    Constraint.name (or ``c{idx}`` when anonymous), using the evaluator's
+    per-source-constraint flat sizes as the source of truth for layout.
+    Returns ``None`` if the input is missing or empty.
+    """
+    if mult_g is None or len(mult_g) == 0:
+        return None
+    out: dict[str, np.ndarray] = {}
+    offset = 0
+    for idx, (c, sz) in enumerate(
+        zip(evaluator._source_constraints, evaluator._constraint_flat_sizes)
+    ):
+        sz = int(sz)
+        chunk = np.asarray(mult_g[offset : offset + sz], dtype=float)
+        key = c.name if c.name else f"c{idx}"
+        out[key] = chunk if sz > 1 else chunk.reshape(())
+        offset += sz
+    if offset != len(mult_g):
+        return None
+    return out
+
+
+def _unpack_bound_duals(
+    model: Model, mult_x: Optional[np.ndarray]
+) -> Optional[dict[str, np.ndarray]]:
+    """Slice a flat bound-multiplier vector into a dict keyed by Variable.name.
+    Returns ``None`` if the input is missing or empty.
+    """
+    if mult_x is None or len(mult_x) == 0:
+        return None
+    out: dict[str, np.ndarray] = {}
+    offset = 0
+    for v in model._variables:
+        size = v.size
+        chunk = np.asarray(mult_x[offset : offset + size], dtype=float)
+        if v.shape == () or v.shape == (1,):
+            out[v.name] = chunk.reshape(v.shape) if v.shape == () else chunk
+        else:
+            out[v.name] = chunk.reshape(v.shape)
+        offset += size
+    if offset != len(mult_x):
+        return None
+    return out
+
+
 def _strong_branch_lp(
     evaluator,
     solution: np.ndarray,
@@ -2578,12 +2626,19 @@ def _solve_continuous(
     if obj_val is not None and model._objective.sense == ObjectiveSense.MAXIMIZE:
         obj_val = -obj_val
 
+    constraint_duals = _unpack_constraint_duals(evaluator, nlp_result.multipliers)
+    bound_duals_lower = _unpack_bound_duals(model, nlp_result.bound_multipliers_lower)
+    bound_duals_upper = _unpack_bound_duals(model, nlp_result.bound_multipliers_upper)
+
     return SolveResult(
         status=status,
         objective=obj_val,
         bound=obj_val if status == "optimal" else None,
         gap=0.0 if status == "optimal" else None,
         x=x_dict,
+        constraint_duals=constraint_duals,
+        bound_duals_lower=bound_duals_lower,
+        bound_duals_upper=bound_duals_upper,
         wall_time=wall_time,
         node_count=0,
         rust_time=0.0,

--- a/python/discopt/solvers/__init__.py
+++ b/python/discopt/solvers/__init__.py
@@ -18,12 +18,19 @@ class SolveStatus(Enum):
 
 @dataclass
 class LPResult:
-    """Result of solving a linear program."""
+    """Result of solving a linear program.
+
+    ``dual_values`` are constraint marginals (one per row).
+    ``reduced_costs`` are variable marginals (one per column). Both are in
+    the sign convention of the LP as passed to the solver (i.e. the
+    internal minimization form).
+    """
 
     status: SolveStatus
     x: Optional[np.ndarray] = None
     objective: Optional[float] = None
     dual_values: Optional[np.ndarray] = None
+    reduced_costs: Optional[np.ndarray] = None
     basis: Optional[object] = None
     iterations: int = 0
     wall_time: float = 0.0
@@ -57,11 +64,21 @@ class QPResult:
 
 @dataclass
 class NLPResult:
-    """Result of solving a nonlinear program."""
+    """Result of solving a nonlinear program.
+
+    ``multipliers`` are constraint Lagrange multipliers (one per constraint
+    row). ``bound_multipliers_lower`` and ``bound_multipliers_upper`` are
+    the multipliers on the variable lower- and upper-bound constraints
+    (one per variable, ≥ 0 at active bounds, ≈ 0 elsewhere). All are in
+    the sign convention of the problem as passed to the solver
+    (i.e. the internal minimization form).
+    """
 
     status: SolveStatus
     x: Optional[np.ndarray] = None
     objective: Optional[float] = None
-    multipliers: Optional[np.ndarray] = None  # constraint multipliers
+    multipliers: Optional[np.ndarray] = None
+    bound_multipliers_lower: Optional[np.ndarray] = None
+    bound_multipliers_upper: Optional[np.ndarray] = None
     iterations: int = 0
     wall_time: float = 0.0

--- a/python/discopt/solvers/lp_highs.py
+++ b/python/discopt/solvers/lp_highs.py
@@ -202,12 +202,15 @@ def solve_lp(
         x = np.array(sol.col_value, dtype=np.float64)
         _, obj = h.getInfoValue("objective_function_value")
         dual = np.array(sol.row_dual, dtype=np.float64)
+        col_dual_raw = np.array(getattr(sol, "col_dual", []), dtype=np.float64)
+        col_dual = col_dual_raw if col_dual_raw.size else None
         basis = h.getBasis()
         return LPResult(
             status=status,
             x=x,
             objective=obj,
             dual_values=dual,
+            reduced_costs=col_dual,
             basis=basis,
             iterations=int(iters),
             wall_time=float(wall_time),

--- a/python/discopt/solvers/nlp_ipopt.py
+++ b/python/discopt/solvers/nlp_ipopt.py
@@ -227,12 +227,20 @@ def solve_nlp(
     multipliers = info.get("mult_g", None)
     if multipliers is not None and len(multipliers) == 0:
         multipliers = None
+    mult_x_L = info.get("mult_x_L", None)
+    if mult_x_L is not None and len(mult_x_L) == 0:
+        mult_x_L = None
+    mult_x_U = info.get("mult_x_U", None)
+    if mult_x_U is not None and len(mult_x_U) == 0:
+        mult_x_U = None
 
     return NLPResult(
         status=status,
         x=np.asarray(x),
         objective=float(info["obj_val"]),
         multipliers=np.asarray(multipliers) if multipliers is not None else None,
+        bound_multipliers_lower=np.asarray(mult_x_L) if mult_x_L is not None else None,
+        bound_multipliers_upper=np.asarray(mult_x_U) if mult_x_U is not None else None,
         iterations=0,  # Ipopt doesn't expose iteration count via this API
         wall_time=wall_time,
     )

--- a/python/discopt/validation/examiner.py
+++ b/python/discopt/validation/examiner.py
@@ -70,6 +70,7 @@ class ExaminerReport:
     n_active_bounds: int = 0
     duals_recovered: bool = False
     dual_recovery_residual: float = 0.0
+    solver_duals_used: bool = False
 
     @property
     def passed(self) -> bool:
@@ -225,11 +226,45 @@ def examine(
             )
         )
 
-    # ── 5-7. Dual-side KKT checks via active-set recovery ───────────────────
+    # ── 5-7. Dual-side KKT checks ───────────────────────────────────────────
+    # Prefer solver-supplied duals when present (true Examiner-style direct
+    # check). Fall back to active-set recovery when the solver does not
+    # expose multipliers. When both are available, also report a cross-check
+    # of recovered-vs-solver multipliers as an extra consistency line.
     n_active_cons = 0
     n_active_bounds = 0
     duals_recovered = False
     dual_residual = 0.0
+    solver_duals_used = False
+
+    solver_dual_vec, solver_lb_vec, solver_ub_vec = _flatten_solver_duals(result, model, evaluator)
+    have_solver_duals = (
+        solver_dual_vec is not None or solver_lb_vec is not None or solver_ub_vec is not None
+    )
+
+    if have_solver_duals and x_flat.size:
+        solver_checks = _check_solver_duals(
+            evaluator=evaluator,
+            x_flat=x_flat,
+            lb=lb,
+            ub=ub,
+            body=body,
+            sense_arr=sense_arr,
+            rhs_arr=rhs_arr,
+            jac=jac,
+            row_labels=row_labels,
+            var_names=var_names,
+            is_continuous=is_continuous,
+            mu=solver_dual_vec,
+            lam_lb=solver_lb_vec,
+            lam_ub=solver_ub_vec,
+            dual_feas_tol=dual_feas_tol,
+            primal_cs_tol=primal_cs_tol,
+            dual_cs_tol=dual_cs_tol,
+        )
+        checks.extend(solver_checks)
+        solver_duals_used = True
+
     if recover_duals and x_flat.size:
         kkt = _recover_and_check_kkt(
             evaluator=evaluator,
@@ -248,12 +283,28 @@ def examine(
             primal_cs_tol=primal_cs_tol,
             dual_cs_tol=dual_cs_tol,
             sense_max=obj_sense,
+            tag="recovered" if solver_duals_used else "",
         )
         checks.extend(kkt["checks"])
         n_active_cons = kkt["n_active_cons"]
         n_active_bounds = kkt["n_active_bounds"]
         duals_recovered = kkt["recovered"]
         dual_residual = kkt["residual"]
+
+        if solver_duals_used and kkt["recovered"]:
+            cons_check = _check_dual_consistency(
+                solver_mu=solver_dual_vec,
+                solver_lb=solver_lb_vec,
+                solver_ub=solver_ub_vec,
+                recovered_mu_full=kkt.get("recovered_mu_full"),
+                recovered_lam_lb_full=kkt.get("recovered_lam_lb_full"),
+                recovered_lam_ub_full=kkt.get("recovered_lam_ub_full"),
+                row_labels=row_labels,
+                var_names=var_names,
+                tol=max(dual_feas_tol * 1e3, 1e-3),
+            )
+            if cons_check is not None:
+                checks.append(cons_check)
 
     merit = float(sum(c.norm2_violation**2 for c in checks))
     return ExaminerReport(
@@ -263,6 +314,7 @@ def examine(
         n_active_bounds=n_active_bounds,
         duals_recovered=duals_recovered,
         dual_recovery_residual=dual_residual,
+        solver_duals_used=solver_duals_used,
     )
 
 
@@ -302,6 +354,217 @@ def _row_metadata(evaluator):
         else:
             labels.append(str(cname))
     return np.asarray(senses), np.asarray(rhss, dtype=float), labels
+
+
+def _flatten_solver_duals(result, model: Model, evaluator):
+    """Flatten ``SolveResult.constraint_duals`` / ``bound_duals_*`` back into
+    vectors aligned with ``evaluator``'s constraint and variable ordering.
+    Returns ``(mu, lam_lb, lam_ub)``; any of the three may be ``None``.
+    """
+
+    def _flatten_named(named, key_iter, sizes):
+        if named is None:
+            return None
+        parts: list[np.ndarray] = []
+        for key, sz in zip(key_iter, sizes):
+            if key not in named:
+                return None
+            arr = np.asarray(named[key], dtype=float).reshape(-1)
+            if arr.size != int(sz):
+                return None
+            parts.append(arr)
+        return np.concatenate(parts) if parts else np.empty(0, dtype=float)
+
+    mu = None
+    if getattr(result, "constraint_duals", None) is not None:
+        keys = [
+            (c.name if c.name else f"c{idx}") for idx, c in enumerate(evaluator._source_constraints)
+        ]
+        mu = _flatten_named(result.constraint_duals, keys, evaluator._constraint_flat_sizes)
+
+    var_keys = [v.name for v in model._variables]
+    var_sizes = [int(v.size) for v in model._variables]
+
+    lam_lb = None
+    if getattr(result, "bound_duals_lower", None) is not None:
+        lam_lb = _flatten_named(result.bound_duals_lower, var_keys, var_sizes)
+    lam_ub = None
+    if getattr(result, "bound_duals_upper", None) is not None:
+        lam_ub = _flatten_named(result.bound_duals_upper, var_keys, var_sizes)
+
+    return mu, lam_lb, lam_ub
+
+
+def _check_dual_consistency(
+    *,
+    solver_mu: Optional[np.ndarray],
+    solver_lb: Optional[np.ndarray],
+    solver_ub: Optional[np.ndarray],
+    recovered_mu_full: Optional[np.ndarray],
+    recovered_lam_lb_full: Optional[np.ndarray],
+    recovered_lam_ub_full: Optional[np.ndarray],
+    row_labels: list[str],
+    var_names: list[str],
+    tol: float,
+) -> Optional[CheckResult]:
+    """Compare solver-supplied vs LSQ-recovered multipliers.
+
+    The recovered multipliers are an LSQ fit, not a solver-quality solution,
+    so we use a permissive tolerance and report the largest absolute discrepancy.
+    Returns ``None`` when there is nothing to compare.
+    """
+    diffs: list[tuple[str, float]] = []
+    if solver_mu is not None and recovered_mu_full is not None and solver_mu.size:
+        d = np.abs(solver_mu - recovered_mu_full)
+        for i, val in enumerate(d):
+            label = row_labels[i] if i < len(row_labels) else f"row{i}"
+            diffs.append((f"μ[{label}]", float(val)))
+    if solver_lb is not None and recovered_lam_lb_full is not None and solver_lb.size:
+        d = np.abs(solver_lb - recovered_lam_lb_full)
+        for i, val in enumerate(d):
+            label = var_names[i] if i < len(var_names) else f"x{i}"
+            diffs.append((f"λ_lb[{label}]", float(val)))
+    if solver_ub is not None and recovered_lam_ub_full is not None and solver_ub.size:
+        d = np.abs(solver_ub - recovered_lam_ub_full)
+        for i, val in enumerate(d):
+            label = var_names[i] if i < len(var_names) else f"x{i}"
+            diffs.append((f"λ_ub[{label}]", float(val)))
+
+    if not diffs:
+        return None
+
+    max_idx = int(np.argmax([v for _, v in diffs]))
+    max_label, max_val = diffs[max_idx]
+    norm2 = float(np.linalg.norm([v for _, v in diffs]))
+    violators = sorted(((lbl, v) for lbl, v in diffs if v > tol), key=lambda kv: -kv[1])[:8]
+    return CheckResult(
+        name="dual_consistency (solver vs recovered)",
+        passed=max_val <= tol,
+        tolerance=tol,
+        max_violation=max_val,
+        norm2_violation=norm2,
+        worst_label=max_label,
+        detail="|solver_dual − recovered_dual|; permissive tol since recovery is LSQ",
+        violators=violators,
+    )
+
+
+def _check_solver_duals(
+    *,
+    evaluator,
+    x_flat: np.ndarray,
+    lb: np.ndarray,
+    ub: np.ndarray,
+    body: np.ndarray,
+    sense_arr: np.ndarray,
+    rhs_arr: np.ndarray,
+    jac: np.ndarray,
+    row_labels: list[str],
+    var_names: list[str],
+    is_continuous: np.ndarray,
+    mu: Optional[np.ndarray],
+    lam_lb: Optional[np.ndarray],
+    lam_ub: Optional[np.ndarray],
+    dual_feas_tol: float,
+    primal_cs_tol: float,
+    dual_cs_tol: float,
+) -> list[CheckResult]:
+    """Run the four dual-side KKT checks using multipliers reported by the
+    solver (Examiner-style direct check, no LSQ recovery).
+
+    Sign convention (cyipopt-compatible, matching how discopt hands the
+    problem to the solver):
+      - "<=" rows have ``cu = 0`` active so ``mu >= 0``
+      - ">=" rows have ``cl = 0`` active so ``mu <= 0``
+      - "==" rows: ``mu`` free
+      - ``lam_lb``, ``lam_ub`` are non-negative (active at the bound)
+    Stationarity in the internal-min form: ``∇f + Jᵀ μ + λ_ub - λ_lb = 0``,
+    where ``∇f`` is what ``evaluator.evaluate_gradient`` returns (already
+    negated for maximize).
+    """
+    cont_idx = np.where(is_continuous)[0]
+    grad = evaluator.evaluate_gradient(x_flat)
+    grad_c = grad[cont_idx]
+    jac_c = jac[:, cont_idx] if jac.size else np.empty((0, cont_idx.size))
+
+    mu_eff = mu if mu is not None else np.zeros(jac.shape[0] if jac.size else 0)
+    lam_lb_full = lam_lb if lam_lb is not None else np.zeros(x_flat.size)
+    lam_ub_full = lam_ub if lam_ub is not None else np.zeros(x_flat.size)
+    lam_lb_c = lam_lb_full[cont_idx]
+    lam_ub_c = lam_ub_full[cont_idx]
+
+    stat = grad_c.copy()
+    if mu_eff.size:
+        stat = stat + jac_c.T @ mu_eff
+    stat = stat + lam_ub_c - lam_lb_c
+    stat_max = float(np.max(np.abs(stat))) if stat.size else 0.0
+    stat_norm = float(np.linalg.norm(stat))
+    worst_var = int(np.argmax(np.abs(stat))) if stat.size else 0
+    cont_names = [var_names[i] for i in cont_idx]
+    stationarity = CheckResult(
+        name="stationarity (solver duals)",
+        passed=stat_max <= dual_feas_tol,
+        tolerance=dual_feas_tol,
+        max_violation=stat_max,
+        norm2_violation=stat_norm,
+        worst_label=cont_names[worst_var] if cont_names else "",
+        detail="∇f + Jᵀμ + λ_ub - λ_lb at solver-supplied duals",
+    )
+
+    sign_viol = np.zeros(mu_eff.size)
+    if mu_eff.size:
+        is_le = sense_arr == "<="
+        is_ge = sense_arr == ">="
+        sign_viol[is_le] = np.maximum(-mu_eff[is_le], 0.0)
+        sign_viol[is_ge] = np.maximum(mu_eff[is_ge], 0.0)
+    sign_max = float(np.max(sign_viol)) if sign_viol.size else 0.0
+    sign_norm = float(np.linalg.norm(sign_viol))
+    worst_row = int(np.argmax(sign_viol)) if sign_viol.size else 0
+    dual_var_feas = CheckResult(
+        name="dual_var_feas (solver duals)",
+        passed=sign_max <= dual_feas_tol,
+        tolerance=dual_feas_tol,
+        max_violation=sign_max,
+        norm2_violation=sign_norm,
+        worst_label=row_labels[worst_row] if row_labels and sign_viol.size else "",
+        detail="μ sign per row sense; λ_lb, λ_ub ≥ 0",
+    )
+
+    cs_p_lb = lam_lb_full * np.where(np.isfinite(lb), np.abs(x_flat - lb), 0.0)
+    cs_p_ub = lam_ub_full * np.where(np.isfinite(ub), np.abs(ub - x_flat), 0.0)
+    cs_p = np.maximum(cs_p_lb, cs_p_ub)
+    cs_p_max = float(np.max(cs_p)) if cs_p.size else 0.0
+    cs_p_norm = float(np.linalg.norm(cs_p))
+    worst_col = int(np.argmax(cs_p)) if cs_p.size else 0
+    primal_cs = CheckResult(
+        name="primal_cs (solver duals)",
+        passed=cs_p_max <= primal_cs_tol,
+        tolerance=primal_cs_tol,
+        max_violation=cs_p_max,
+        norm2_violation=cs_p_norm,
+        worst_label=var_names[worst_col] if var_names and cs_p.size else "",
+        detail="λ · |x − bound| over all variables",
+    )
+
+    if mu_eff.size:
+        signed = body - rhs_arr
+        cs_d = np.abs(mu_eff) * np.abs(signed)
+        cs_d_max = float(np.max(cs_d))
+        cs_d_norm = float(np.linalg.norm(cs_d))
+        worst_row2 = int(np.argmax(cs_d))
+        dual_cs = CheckResult(
+            name="dual_cs (solver duals)",
+            passed=cs_d_max <= dual_cs_tol,
+            tolerance=dual_cs_tol,
+            max_violation=cs_d_max,
+            norm2_violation=cs_d_norm,
+            worst_label=row_labels[worst_row2] if row_labels else "",
+            detail="|μ| · |body − rhs| over all rows",
+        )
+    else:
+        dual_cs = CheckResult(name="dual_cs (solver duals)", passed=True, tolerance=dual_cs_tol)
+
+    return [stationarity, dual_var_feas, primal_cs, dual_cs]
 
 
 def _check_var_bounds(
@@ -413,6 +676,7 @@ def _recover_and_check_kkt(
     primal_cs_tol: float,
     dual_cs_tol: float,
     sense_max: ObjectiveSense,
+    tag: str = "",
 ) -> dict:
     """Recover multipliers from the active set and run dual-side KKT checks.
 
@@ -422,6 +686,7 @@ def _recover_and_check_kkt(
     Examiner fixes integers and runs continuous KKT, which is exactly what
     omitting those columns does.
     """
+    suffix = f" ({tag})" if tag else ""
     grad = evaluator.evaluate_gradient(x_flat)
     cont_idx = np.where(is_continuous)[0]
     if cont_idx.size == 0:
@@ -468,7 +733,7 @@ def _recover_and_check_kkt(
         return {
             "checks": [
                 CheckResult(
-                    name="stationarity",
+                    name=f"stationarity{suffix}",
                     passed=max_r <= dual_feas_tol,
                     tolerance=dual_feas_tol,
                     max_violation=max_r,
@@ -481,6 +746,9 @@ def _recover_and_check_kkt(
             "n_active_bounds": 0,
             "recovered": True,
             "residual": norm_r,
+            "recovered_mu_full": np.zeros(jac.shape[0]) if jac.size else np.zeros(0),
+            "recovered_lam_lb_full": np.zeros(x_flat.size),
+            "recovered_lam_ub_full": np.zeros(x_flat.size),
         }
 
     cols = []
@@ -520,14 +788,32 @@ def _recover_and_check_kkt(
         msg = f"dual recovery failed: {e}"
         return {
             "checks": [
-                CheckResult(name="stationarity", passed=False, tolerance=dual_feas_tol, detail=msg),
-                CheckResult(name="primal_cs", passed=False, tolerance=primal_cs_tol, detail=msg),
-                CheckResult(name="dual_cs", passed=False, tolerance=dual_cs_tol, detail=msg),
+                CheckResult(
+                    name=f"stationarity{suffix}",
+                    passed=False,
+                    tolerance=dual_feas_tol,
+                    detail=msg,
+                ),
+                CheckResult(
+                    name=f"primal_cs{suffix}",
+                    passed=False,
+                    tolerance=primal_cs_tol,
+                    detail=msg,
+                ),
+                CheckResult(
+                    name=f"dual_cs{suffix}",
+                    passed=False,
+                    tolerance=dual_cs_tol,
+                    detail=msg,
+                ),
             ],
             "n_active_cons": int(n_mu),
             "n_active_bounds": int(n_llb + n_lub),
             "recovered": False,
             "residual": float("inf"),
+            "recovered_mu_full": None,
+            "recovered_lam_lb_full": None,
+            "recovered_lam_ub_full": None,
         }
 
     stat_resid = A @ y - b
@@ -535,7 +821,7 @@ def _recover_and_check_kkt(
     stat_norm = float(np.linalg.norm(stat_resid))
     worst_var = int(np.argmax(np.abs(stat_resid))) if stat_resid.size else 0
     stationarity = CheckResult(
-        name="stationarity",
+        name=f"stationarity{suffix}",
         passed=stat_max <= dual_feas_tol,
         tolerance=dual_feas_tol,
         max_violation=stat_max,
@@ -559,7 +845,7 @@ def _recover_and_check_kkt(
     cs_p_max = float(np.max(cs_p)) if cs_p.size else 0.0
     cs_p_norm = float(np.linalg.norm(cs_p))
     primal_cs = CheckResult(
-        name="primal_cs",
+        name=f"primal_cs{suffix}",
         passed=cs_p_max <= primal_cs_tol,
         tolerance=primal_cs_tol,
         max_violation=cs_p_max,
@@ -575,7 +861,7 @@ def _recover_and_check_kkt(
         cs_d_norm = float(np.linalg.norm(cs_d))
         worst_row = int(np.argmax(cs_d))
         dual_cs = CheckResult(
-            name="dual_cs",
+            name=f"dual_cs{suffix}",
             passed=cs_d_max <= dual_cs_tol,
             tolerance=dual_cs_tol,
             max_violation=cs_d_max,
@@ -584,7 +870,24 @@ def _recover_and_check_kkt(
             detail="max |μ|·|body−rhs| over recovered active rows",
         )
     else:
-        dual_cs = CheckResult(name="dual_cs", passed=True, tolerance=dual_cs_tol)
+        dual_cs = CheckResult(name=f"dual_cs{suffix}", passed=True, tolerance=dual_cs_tol)
+
+    # Pack recovered duals into full-length vectors aligned with the
+    # evaluator's row order and the model's variable order, so the cross-check
+    # against solver-supplied duals can compare element-wise.
+    mu_full = np.zeros(jac.shape[0]) if jac.size else np.zeros(0)
+    if n_mu:
+        # mu_act is in "flipped to ≤ form"; un-flip ">=" rows back to original sign.
+        sub_sense = sense_arr[row_select]
+        mu_signed = mu.copy()
+        mu_signed[sub_sense == ">="] *= -1.0
+        mu_full[row_select] = mu_signed
+    lam_lb_full = np.zeros(x_flat.size)
+    lam_ub_full = np.zeros(x_flat.size)
+    for k, j in enumerate(lb_active):
+        lam_lb_full[cont_idx[j]] = lam_lb[k]
+    for k, j in enumerate(ub_active):
+        lam_ub_full[cont_idx[j]] = lam_ub[k]
 
     return {
         "checks": [stationarity, primal_cs, dual_cs],
@@ -592,4 +895,7 @@ def _recover_and_check_kkt(
         "n_active_bounds": int(n_llb + n_lub),
         "recovered": True,
         "residual": stat_norm,
+        "recovered_mu_full": mu_full,
+        "recovered_lam_lb_full": lam_lb_full,
+        "recovered_lam_ub_full": lam_ub_full,
     }

--- a/python/tests/test_examiner.py
+++ b/python/tests/test_examiner.py
@@ -127,3 +127,202 @@ def test_recover_duals_can_be_disabled():
     rep = examine(res, m, recover_duals=False)
     assert rep.passed
     assert not any(c.name == "stationarity" for c in rep.checks)
+
+
+# ─────────────────────────────────────────────────────────────────────────────
+# Solver-supplied dual path
+# ─────────────────────────────────────────────────────────────────────────────
+
+
+def _result_with_duals(
+    x: dict[str, np.ndarray],
+    obj: float | None,
+    *,
+    constraint_duals=None,
+    bound_duals_lower=None,
+    bound_duals_upper=None,
+) -> SolveResult:
+    return SolveResult(
+        status="optimal",
+        objective=obj,
+        x=x,
+        constraint_duals=constraint_duals,
+        bound_duals_lower=bound_duals_lower,
+        bound_duals_upper=bound_duals_upper,
+    )
+
+
+def test_solver_duals_pass_at_correct_kkt_point():
+    """KKT for min (x-2)^2+(y+1)^2 s.t. x+y<=1, y>=0 at (1,0).
+
+    Stationarity:
+      ∇f = (2(x-2), 2(y+1)) = (-2, 2).
+      Active: row "x+y<=1" (row index 0), bound y>=0 (lb on y).
+      ∇f + Jᵀ μ + λ_ub − λ_lb = 0
+      → (-2, 2) + (1, 1)·μ + (0, -λ_lb_y) = 0
+      → μ = 2, λ_lb_y = 4.
+    """
+    m = dm.Model("kkt_demo")
+    x = m.continuous("x", lb=0.0, ub=5.0)
+    y = m.continuous("y", lb=0.0, ub=5.0)
+    m.minimize((x - 2) ** 2 + (y + 1) ** 2)
+    m.subject_to(x + y <= 1, name="c0")
+
+    res = _result_with_duals(
+        {"x": np.array(1.0), "y": np.array(0.0)},
+        obj=2.0,
+        constraint_duals={"c0": np.array(2.0)},
+        bound_duals_lower={"x": np.array(0.0), "y": np.array(4.0)},
+        bound_duals_upper={"x": np.array(0.0), "y": np.array(0.0)},
+    )
+    rep = examine(res, m)
+    assert rep.passed, rep.summary(verbose=True)
+    assert rep.solver_duals_used
+    assert any(c.name == "stationarity (solver duals)" for c in rep.checks)
+    assert any(c.name == "stationarity (recovered)" for c in rep.checks)
+
+
+def test_solver_duals_flag_wrong_sign():
+    """A "<=" row supplied with negative μ violates dual feasibility."""
+    m = dm.Model("wrong_sign")
+    x = m.continuous("x", lb=0.0, ub=5.0)
+    y = m.continuous("y", lb=0.0, ub=5.0)
+    m.minimize((x - 2) ** 2 + (y + 1) ** 2)
+    m.subject_to(x + y <= 1, name="c0")
+
+    res = _result_with_duals(
+        {"x": np.array(1.0), "y": np.array(0.0)},
+        obj=2.0,
+        constraint_duals={"c0": np.array(-2.0)},  # wrong sign for "<="
+        bound_duals_lower={"x": np.array(0.0), "y": np.array(4.0)},
+        bound_duals_upper={"x": np.array(0.0), "y": np.array(0.0)},
+    )
+    rep = examine(res, m)
+    bad = next(c for c in rep.checks if c.name == "dual_var_feas (solver duals)")
+    assert not bad.passed
+    assert bad.max_violation > DUAL_FEAS_TOL
+
+
+def test_solver_duals_flag_stationarity_when_inconsistent():
+    """Supply duals that don't satisfy ∇f + Jᵀμ + λ_ub − λ_lb = 0."""
+    m = dm.Model("bad_stat")
+    x = m.continuous("x", lb=0.0, ub=5.0)
+    y = m.continuous("y", lb=0.0, ub=5.0)
+    m.minimize((x - 2) ** 2 + (y + 1) ** 2)
+    m.subject_to(x + y <= 1, name="c0")
+
+    res = _result_with_duals(
+        {"x": np.array(1.0), "y": np.array(0.0)},
+        obj=2.0,
+        constraint_duals={"c0": np.array(0.5)},  # too small
+        bound_duals_lower={"x": np.array(0.0), "y": np.array(0.5)},
+        bound_duals_upper={"x": np.array(0.0), "y": np.array(0.0)},
+    )
+    rep = examine(res, m)
+    bad = next(c for c in rep.checks if c.name == "stationarity (solver duals)")
+    assert not bad.passed
+    assert bad.max_violation > DUAL_FEAS_TOL
+
+
+def test_solver_duals_flag_dual_cs_for_inactive_row():
+    """An inactive "<=" row with nonzero μ violates dual complementary slack."""
+    m = dm.Model("dual_cs")
+    x = m.continuous("x", lb=-10.0, ub=10.0)
+    m.minimize((x - 2) ** 2)
+    m.subject_to(x <= 5, name="c0")  # inactive at x=2
+
+    res = _result_with_duals(
+        {"x": np.array(2.0)},
+        obj=0.0,
+        constraint_duals={"c0": np.array(1.0)},  # μ ≠ 0 but body-rhs = -3
+        bound_duals_lower={"x": np.array(0.0)},
+        bound_duals_upper={"x": np.array(0.0)},
+    )
+    rep = examine(res, m)
+    bad = next(c for c in rep.checks if c.name == "dual_cs (solver duals)")
+    assert not bad.passed
+    assert bad.max_violation > 1e-6
+
+
+def test_dual_consistency_check_appears_when_both_paths_run():
+    m = dm.Model("xcheck")
+    x = m.continuous("x", lb=0.0, ub=5.0)
+    y = m.continuous("y", lb=0.0, ub=5.0)
+    m.minimize((x - 2) ** 2 + (y + 1) ** 2)
+    m.subject_to(x + y <= 1, name="c0")
+
+    res = _result_with_duals(
+        {"x": np.array(1.0), "y": np.array(0.0)},
+        obj=2.0,
+        constraint_duals={"c0": np.array(2.0)},
+        bound_duals_lower={"x": np.array(0.0), "y": np.array(4.0)},
+        bound_duals_upper={"x": np.array(0.0), "y": np.array(0.0)},
+    )
+    rep = examine(res, m)
+    cross = next(
+        (c for c in rep.checks if c.name == "dual_consistency (solver vs recovered)"),
+        None,
+    )
+    assert cross is not None
+    assert cross.passed, cross.detail
+
+
+def test_end_to_end_nlp_duals_round_trip_through_examine():
+    """Solve a small NLP via discopt; confirm duals populated and pass examine()."""
+    pytest.importorskip("cyipopt")
+
+    m = dm.Model("nlp_xyz")
+    x = m.continuous("x", lb=0.1, ub=5.0)
+    y = m.continuous("y", lb=0.1, ub=5.0)
+    m.minimize(dm.exp(x - 2) + dm.log(y + 1))
+    m.subject_to(x + y <= 1, name="c0")
+
+    res = m.solve()
+    assert res.status == "optimal"
+    # NLP path should populate solver-supplied duals.
+    assert res.constraint_duals is not None
+    assert res.bound_duals_lower is not None
+    assert res.bound_duals_upper is not None
+    assert "c0" in res.constraint_duals
+    assert "x" in res.bound_duals_lower
+    assert "y" in res.bound_duals_lower
+
+    rep = examine(res, m)
+    assert rep.solver_duals_used
+    # All KKT checks should pass at the IPOPT solution.
+    assert rep.passed, rep.summary(verbose=True)
+    # Both solver-direct and recovered checks should be present.
+    names = {c.name for c in rep.checks}
+    assert "stationarity (solver duals)" in names
+    assert "stationarity (recovered)" in names
+    # Cross-check should appear and pass (within permissive tol).
+    assert "dual_consistency (solver vs recovered)" in names
+
+
+def test_dual_consistency_check_flags_disagreement():
+    """Solver duals that solve KKT but differ wildly from a recovered LSQ
+    solution should still trip the consistency cross-check (large tol margin).
+    """
+    m = dm.Model("disagree")
+    # Underdetermined active set: two equality redundant variants → many y solve KKT.
+    x = m.continuous("x", lb=0.0, ub=5.0)
+    y = m.continuous("y", lb=0.0, ub=5.0)
+    m.minimize((x - 2) ** 2 + (y + 1) ** 2)
+    m.subject_to(x + y <= 1, name="c0")
+
+    # supply hugely off-magnitude (but still "satisfies" inequality sign) duals
+    # by combining with a wrong stationarity offset injected via lb/ub fields
+    res = _result_with_duals(
+        {"x": np.array(1.0), "y": np.array(0.0)},
+        obj=2.0,
+        constraint_duals={"c0": np.array(2.0)},
+        bound_duals_lower={"x": np.array(50.0), "y": np.array(54.0)},
+        bound_duals_upper={"x": np.array(0.0), "y": np.array(0.0)},
+    )
+    rep = examine(res, m)
+    cross = next(
+        (c for c in rep.checks if c.name == "dual_consistency (solver vs recovered)"),
+        None,
+    )
+    assert cross is not None
+    assert not cross.passed

